### PR TITLE
[Tranaform Dialect] Add new iree.compute_workgroup_count_region opera…

### PR DESCRIPTION
**Pull request by Sameeranjoshi**
…tion.

This patch carves out the code from 'TileToForallAndWorkgroupCountRegionOp', and adds a new operation in transform dialect called 'iree.compute_workgroup_count_region'. This was needed to support some custom pipeline which had a need to seperate this operation.

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Could you provide some more rationale for this. I'm  currently changing the workgroup count computation handling to something more general. So want to understand the need here

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
(see comment for request for clarification)

**Comment by Sameeranjoshi** [comment]
Thanks for looking at it, we are trying to do experiments with more than 3-level tiling . 
The combined approach in `TileToForallAndWorkgroupCountRegionOp` restricts to 3 level tile sizes, we need more level to specify in the `tile_sizes` attribute, we are doing this with a few more supported patches in mlir core (https://reviews.llvm.org/D147830#change-Z5DR4wmNOs1v)]

**Comment by Sameeranjoshi** [comment]
A gentle reminder ping.
Thanks

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    }
+  }
+}
+
+transform.sequence failures(propagate) {
+^bb1(%variant_op: !pdl.operation):
+  %original_matmul = transform.structured.match ops{["linalg.matmul"]} in %variant_op
+    : (!pdl.operation) -> !pdl.operation
+
+  transform.iree.compute_workgroup_count_region %original_matmul
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir:45]: I am not sure this is always going to work.... you also need to ensure that the corresponding tiling is also consistent with this (AFAIK based on how the `scf.for_all` is normalized

**Comment by MaheshRavishankar** [comment]
> A gentle reminder ping. Thanks

I am not fully following what the ask here is (maybe more concrete description of what is being tried would help.) In any case, I am in the process of reworking the way workgroup count region is handled in #13038 . I am not sure that helps (I have to fix transform dialect use case on that PR). So some more details would help square these.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir (hunk 1 above)
Comment (Sameeranjoshi) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir:45]: Yes, I think I needed a test like

```
%0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
%1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
transform.iree.compute_workgroup_count_region %2 num_threads [256]
```

Our use case was to do something like below test case in IREE, but we couldn't do hence we pushed a [patch](https://reviews.llvm.org/D147830#change-Z5DR4wmNOs1v)%5D) in MLIR to support the same and it needs this patch, hence we tried to upstream it, this patch just extracts the logic of workgroup count. 

Does your patch handle the use case we are trying to achieve with coalesce?
```

%0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
%1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
%3 = transform.loop.coalesce_parallel %1 { collapsed_dim0 = array<i64: 0, 1>, mapping = [#gpu.block<x>] } : (!pdl.operation) -> !pdl.operation
%4 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
```

Thanks again for reviewing.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/transform_dialect_iree_compute_workgroup_count_region.mlir:45]: > Yes, I think I needed a test like
> 
> ```
> %0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> %1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
> transform.iree.compute_workgroup_count_region %2 num_threads [256]
> ```
> 
> Our use case was to do something like below test case in IREE, but we couldn't do hence we pushed a [patch](https://reviews.llvm.org/D147830#change-Z5DR4wmNOs1v)%5D) in MLIR to support the same and it needs this patch, hence we tried to upstream it, this patch just extracts the logic of workgroup count.
> 
> Does your patch handle the use case we are trying to achieve with coalesce?
> 
> ```
> 
> %0 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> %1, %2 = transform.structured.tile_to_forall_op %0 num_threads [16, 16]
> %3 = transform.loop.coalesce_parallel %1 { collapsed_dim0 = array<i64: 0, 1>, mapping = [#gpu.block<x>] } : (!pdl.operation) -> !pdl.operation
> %4 = transform.structured.match ops{["linalg.matmul"]} in %variant_op : (!pdl.operation) -> !pdl.operation
> ```
> 
> Thanks again for reviewing.

I think the coalescing should be folded into the `structured.tile_to_forall_op` (thats how I think it is handled today in IREE).
If you can post the IR after `transform.loop.coalese_parallel` I can let you know if it works or not. I suspect it does.

**Comment by Sameeranjoshi** [comment]
@MaheshRavishankar merged another PR for handling in a more sophisticated way. Thanks, closing this.

**Pull request closed without merging**